blog/2024-11-22-inspektor-gadget-unittesting-framework.md

```diff
@@ -0,0 +1,265 @@
+
+Objective: Verify the gadget does not capture events when the filter is configured for a different namespace.
+
+Validation: No events should be detected.
+#### Events Matching Specific Namespace Filters
+
+Objective: Test that events are captured only for processes operating in a matching mount namespace.
+
+Validation: The captured event should include details such as file name, flags, and process metadata.
+#### Testing File Flags and Modes
+
+Objective: Confirm that the gadget accurately captures the flags and modes used when opening a file.
+
+Setup: A file is created with specific access flags and modes.
+
+Validation: The flags and modes in the captured event are compared to the expected values.
+
+### Example Test: Capturing Events with Filters
+Let’s look at a simplified example of one test case:
+```go
+"captures_events_with_matching_filter": {
+    runnerConfig: &utilstest.RunnerConfig{},
+    mntnsFilterMap: func(info *utilstest.RunnerInfo) *ebpf.Map {
+        return utilstest.CreateMntNsFilterMap(t, info.MountNsID)
+    },
+    generateEvent: func() (int, error) {
+        return unix.Open("/dev/null", 0, 0)
+    },
+    validateEvent: func(t *testing.T, info *utilstest.RunnerInfo, fd int, events []ExpectedTraceOpenEvent) {
+        utilstest.ExpectOneEvent(func(info *utilstest.RunnerInfo, fd int) *ExpectedTraceOpenEvent {
+            return &ExpectedTraceOpenEvent{
+                Proc:  info.Proc,
+                Fd:    uint32(fd),
+                FName: "/dev/null",
+            }
+        })(t, info, fd, events)
+    },
+},
+```
+#### How It Works
+- The test configures a namespace filter map for the runner.
+- A dummy event is generated by opening /dev/null.
+- The test validates that the event is captured and matches the expected file name.
+
+### Test Setup
+The structure follows the principles of table-driven testing, allowing us to iterate through multiple test cases with different configurations and expected outcomes. Each test runs independently, enabling parallel execution.
+
+```go
+for name, testCase := range testCases {
+    t.Run(name, func(t *testing.T) {
+        t.Parallel()
+        // ...
+    })
+}
+```
+
+### Initializing the Runner
+A runner is a critical component here, acting as the test environment where the functions generating events will operate. If namespace filtering is required, a mntnsFilterMap (mount namespace filter) is also initialized.
+```go
+runner := utilstest.NewRunnerWithTest(t, testCase.runnerConfig)
+var mntnsFilterMap *ebpf.Map
+if testCase.mntnsFilterMap != nil {
+    mntnsFilterMap = testCase.mntnsFilterMap(runner.Info)
+}
+```
+
+## Defining Event Generating Function's Execution
+
+The unit-testing framework provide 2 options for running the function 
+- `beforeGadgetRun`: This function is executed before the gadget starts running, It can be use to execute something before the gadget is started. This is also useful to test Snapshotter Gadgets.
```

Comment:
```suggestion
- `beforeGadgetRun`: This function is executed before the gadget starts running. It can be use to execute something before the gadget is started. This is also useful to test Snapshotter Gadgets.
```